x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearningUsageTransportAction.java

```diff
@@ -227,6 +268,66 @@ private void addDataFrameAnalyticsUsage(GetDataFrameAnalyticsStatsAction.Respons
         }
     }
 
+    private static void initializeStats(Map<String, Long> emptyStatsMap) {
+        emptyStatsMap.put("sum", 0L);
+        emptyStatsMap.put("min", 0L);
+        emptyStatsMap.put("max", 0L);
+    }
+
+    private static void updateStats(Map<String, Long> statsMap, Long value) {
+        statsMap.compute("sum", (k, v) -> v + value);
+        statsMap.compute("min", (k, v) -> Math.min(v, value));
+        statsMap.compute("max", (k, v) -> Math.max(v, value));
+    }
+
+    //TODO separate out ours and users models possibly regression vs classification
+    private void addTrainedModelStats(SearchResponse response, Map<String, Object> inferenceUsage) {
+        inferenceUsage.put("trained_models",
+            Collections.singletonMap(MachineLearningFeatureSetUsage.ALL, createCountUsageEntry(response.getHits().getTotalHits().value)));
+    }
+
+    //TODO separate out ours and users models possibly regression vs classification
+    private void addInferenceIngestUsage(NodesStatsResponse response, Map<String, Object> inferenceUsage) {
+        Set<String> pipelines = new HashSet<>();
+        Map<String, Long> docCountStats = new HashMap<>(3);
+        Map<String, Long> timeStats = new HashMap<>(3);
+        Map<String, Long> failureStats = new HashMap<>(3);
+        initializeStats(docCountStats);
+        initializeStats(timeStats);
+        initializeStats(failureStats);
+
+        response.getNodes()
+            .stream()
+            .map(NodeStats::getIngestStats)
+            .map(IngestStats::getProcessorStats)
+            .forEach(map ->
+                map.forEach((pipelineId, processors) -> {
+                    boolean containsInference = false;
+                    for(IngestStats.ProcessorStat stats : processors) {
+                        if (stats.getName().equals(InferenceProcessor.TYPE)) {
+                            containsInference = true;
+                            long ingestCount = stats.getStats().getIngestCount();
+                            long ingestTime = stats.getStats().getIngestTimeInMillis();
+                            long failureCount = stats.getStats().getIngestFailedCount();
+                            updateStats(docCountStats, ingestCount);
+                            updateStats(timeStats, ingestTime);
+                            updateStats(failureStats, failureCount);
+                        }
+                    }
+                    if (containsInference) {
+                        pipelines.add(pipelineId);
+                    }
+                })
+            );
+
+        Map<String, Object> ingestUsage = new HashMap<>(4);
+        ingestUsage.put("pipelines", createCountUsageEntry(pipelines.size()));
+        ingestUsage.put("docs", docCountStats);
+        ingestUsage.put("time", timeStats);
+        ingestUsage.put("failures", failureStats);
+        inferenceUsage.put("processors", Collections.singletonMap(MachineLearningFeatureSetUsage.ALL, ingestUsage));
```

Comment:
Maybe spell out more clearly as `ingest_processors`.  We need to think as well about how these stats will evolve once we allow inference in searches (either as pipeline aggregations or on individual documents retrieved).  Would we want to include stats for that?  If so what could the names be?  And would the current structure need changing if search time stats were added?  (I am not saying we need to add these stats now, just that we should think about whether they would fit without changing the structure of the ingest pipeline stats.)

Reply:
`ingest_processors` makes sense. 

> Would we want to include stats for [searches]? If so what could the names be? And would the current structure need changing if search time stats were added?

I am not sure how we would unless we record at search time. Depends on the implementation I suppose. As for the current structure, I don't see any conflict. All these stats are under the `processors` (or `ingest_processors`) field. Clearly indicating that this is ONLY for ingest stats. 

I do agree a larger number of folks should give feedback :).